### DISCUSSION TITLE
Big improvements for Windows users, preview 1.12.8_01

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#2113] [Announcements]
I replaced the launcher by new standard java launcher for windows providing better integration with the task bar.
And since 1.12.7_08 it integrates Java runtime provided by Jetbrains with support for colorful emoji symbols in texts.

I am going to release 1.12.7 or even 1.12.8 soon.

To keep it easy, I am no longer releasing windows installers without Java because they were downloaded only by small minority of users.

**Comment by euu2021** [comment]
> And since 1.12.7_08 it integrates Java runtime provided by Jetbrains with support for colorful emoji symbols in texts.

How to get the colorful emojis? The best I could get was this, using the Dialog font:

![image](https://github.com/user-attachments/assets/b0330c74-4f96-404a-b6d3-1fe599c5b20b)

**Reply by dpolivaev** [reply]
Please upload your mind map and not only the screenshot.

**Reply by euu2021** [reply]
example map: 
[EXAMPLE emojis.zip](https://github.com/user-attachments/files/17791828/EXAMPLE.emojis.zip)

**Reply by dpolivaev** [reply]
![freeplane_w1p842TCj7](https://github.com/user-attachments/assets/f5ffbd55-c928-47f1-884a-815c2a1df16b)
It works only with embedded java, but it works

**Comment by HaaiHenkie** [comment]
Hello Dimitry,

Thanks for the improvements

> To keep it easy, I am no longer releasing windows installers without Java because they were downloaded only by small minority of users.

I prefer the Window installation without Java. I think the reason that it is downloaded by an minority of users is obvious. If you are on Windows, the "Download" button and "Looking for the latest version?" on respectively the SourceForge Summary and Files pages point to the installer with Java. So as a user you have to know that there is an installer without Java and you need 2 or 3 extra clicks to download it. That is automatically a minority of users. Despite this almost 2000 users downloaded Freeplane-Setup-1.12.6.exe since its release on 16 september. So I think there is still a need for the installer without Java.

**Reply by dpolivaev** [reply]
I can not deliver same experience for all versions. The version with the zip file can still have the old launcher, but for people who use the installer for windows having different versions that work differently is too confusing and too difficult to support IMHO.

**Reply by HaaiHenkie** [reply]
OK I understand that the new standard java launcher for windows will not work with every Java version. But is that not a simple instruction like "you should use java version x or higher"? I do not expect that is confusing for people who intentionally want to use the installer without Java, since I expect them to be aware which version of Java they have installed. Or is it more complicated than that?

**Reply by dpolivaev** [reply]
With the new launcher I not do not only use a specific Java implementation by Jetbrains but also modify one of it internal configuration files to avoid exceptions thrown by text editing fields on some characters. I think that doing it on their own is too diifficult for average users.

**Reply by dpolivaev** [reply]
See https://github.com/freeplane/freeplane/issues/1961

**Reply by dpolivaev** [reply]
Also, the new freeplane.exe works only with embedded java and does not allow to use external java. But it provides better integration with windows. Having a distribution without embedded java would require to use the old freeplane.exe there, and then user experience would be different and cause many bug reports and learning problems.

**Reply by HaaiHenkie** [reply]
> With the new launcher I not do not only use a specific Java implementation by Jetbrains but also modify one of it internal configuration files to avoid exceptions thrown by text editing fields on some characters. I think that doing it on their own is too difficult for average users.

Yes you are right, that is not very straightforward.

**Comment by HaaiHenkie** [comment]
I have installed Freeplane preview 1.12.8_01 and could indeed see better integration with the task bar: when you right click the Freeplane icon on the task bar you now get a whole list of options while before the only option was to close the Window.

I wanted to know if the new standard java launcher is part of freeplanelauncher.jar, so I tried to start it though freeplane.bat. To be sure I used the embedded JetBrains Runtime java.exe. (BTW, I also had to add the option `-Dorg.freeplane.main.application.FreeplaneSecurityManager.disable=true`) I did not see the task bar improvement. That means freeplanelauncher.jar still uses the original launcher, correct?

That is good news, because the Freeplane Gradle plugin and scripts launching Freeplane will still work.

Just out of interest, where is the new standard java launcher located? Is it part of freeplane.exe?

Suppose one needs to test or debug Freeplane as launched the new standard java launcher, would that be possible?

**Comment by walmar** [comment]
So far I have only installed the preview in a sandbox in Sandboxie Plus, as I tend to do when testing software. However, since v1.8.6 the application no longer starts correctly this way.
In the past, I kind of worked around this issue by using the portable version with a separate user config directory instead, but that doesn't seems to be an option in this case.

From that point I can already see that the preview contains the new subfolder *runtime*. This takes up over 164 MB of disk space.

At the moment I have installed OpenJDK11U-jre_x86-32_windows_hotspot_11.0.24_8.msi for daily usage.
It:
* claims 100 MB of disk space
* is [required](https://www.openoffice.org/download/common/java.html) for Apache_OpenOffice_4.1.15_Win_x86_install_en_US.exe
   * with over 200k downloads per week on [Sourceforge](https://sourceforge.net/projects/openofficeorg.mirror/) web page
   * that uses [approximately 440 MB](https://www.openoffice.org/dev_docs/source/sys_reqs_aoo41.html) disk space after installation
      * That's about 1 GB less disk space than the bloated [LibreOffice](https://www.libreoffice.org/get-help/system-requirements/) requires.
* meets the [requirements](https://docs.freeplane.org/getting-started/Distributions.html#distributions-for-microsoft-windows) for Freeplane-Setup-1.12.6.exe

Even though I assume that the advantages mentioned are there, I currently doubt that it would be a big improvement for me overall.

I think the best solution to this situation for me and possibly others would be if the announced plans were reconsidered and I could still download the installer without the embedded Java Runtime Environment (JRE).
As for the
* **need for an installer without embedded JRE in the past**

  I assume there are mainly two legitimate reasons to choose this: if multiple Java applications are running with it and to manage the JRE separately from the application.
  On the other side, when installing with embedded JRE, there may not necessarily be legitimate cases where users try out whatever they find interesting without bothering much about a clean maintained system.

* **desired taskbar behavior**

  There are numerous claims on the Stack Overflow website over the past 12 years to questions – as linked in #1882 and [further explained](https://stackoverflow.com/questions/9342651/pinning-a-java-application-using-launch4j-to-the-windows-7-taskbar#9396377) – that the application can be handled properly by Windows if
     1. the [[Icons] section](https://jrsoftware.org/ishelp/topic_iconssection.htm) is defined with AppUserModelID for Inno Setup to create a shortcut in the start menu and
	 2. [SetCurrentProcessExplicitAppUserModelID](https://java-native-access.github.io/jna/5.3.1/javadoc/com/sun/jna/platform/win32/Shell32.html#SetCurrentProcessExplicitAppUserModelID-com.sun.jna.WString-) is called via JNA – with [example](https://stackoverflow.com/questions/1907735/using-jna-to-get-set-application-identifier#38762346) – to identify the application process to the taskbar.
* **colorful emojis symbols support in texts**

  I can not write much about it.
* **Dingbats and Symbols fonts related error prevention**

  I assume that the developers of the JRE still have to fix this issue on their side.
  Until then, the documentation could recommend the discovered configuration approach for the respective installation to work around the known Java font issues.
* **different user experience without embedded JRE**

  The installer may be provided with a different file name such as "minimal", "experimental" or "unstable" and will indicate the different behavior and required settings of the JRE, which the user has to setup on its own based on the information provided and without further assistance.

Despite the newly emerging friction related to the announced plan to release the installer only with embedded JRE, I thank you, dpolivaev, for your continued tremendous work on this unique project.

**Reply by HaaiHenkie** [reply]
In this respect I was wondering the following. The taskbar improvements are typical Windows-only. But the emoji and font related improvements and fixes, shouldn't they be applied to all platforms? For MacOS that would be easy, since it is shipped with a JRE already. But is shipping with JRE acceptable for  Linux. And if not, is it acceptable that Freeplane functionality for Linux will diverge?

**Reply by walmar** [reply]
I extracted the contents of ``freeplane_bin-1.12.8.zip\freeplane-1.12.8\`` to the existing path ``C:\Program Files\Freeplane`` to overwrite the installed files before v1.12.8.
After that, it should still be possible to uninstall the application and remove leftovers from the installation path if necessary.

This approach has the
* disadvantage that the program information is not updated as listed in Windows'  ``appwiz.cpl`` and
* advantage that the following steps do not have to be repeated as an administrator after each update, as was previously the case.
   * Execute a .reg file with the following content in order to add the *Freeplane mind map* item to the File Explorer "New" context menu (similar to [How to Add or Remove Items from “New” Context Menu in Windows – AskVG](https://www.askvg.com/how-to-add-remove-items-from-new-menu-in-windows/)) to create new files from a [template mind map](https://docs.freeplane.org/user-documentation/templates.html).
      ````reg
      Windows Registry Editor Version 5.00
      
      [HKEY_CLASSES_ROOT\.mm\ShellNew]
      "NullFile"=""
      "FileName"="C:\\Users\\YourUserName\\AppData\\Roaming\\Freeplane\\1.12.x\\templates\\YourTemplateFileName.mm"
      
      ````
   * Install the *[IFilterShop FreeMind IFilter](https://web.archive.org/web/20180216012845/http://www.ifiltershop.com/freemind-ifilter.html) Desktop Edition* plug-in (with [additional setup steps](https://web.archive.org/web/20170217044407/http://ifiltershop.com/downloads/freemindfilter/readme.html#_Windows_Desktop_Search), if not already done) that allows Microsoft Search products and services to index maps generated by FreeMind mind-mapping software. And then execute ``services.msc`` to restart the *Windows Search* service.

**Reply by dpolivaev** [reply]
Emojis worked on Mac even before. There is no distribution including Java for Linux, and I do not know what happens there.
